### PR TITLE
Optimize `kronecker_solve`, add `KroneckerLinearSolver`

### PR DESCRIPTION
**Pull request by davschneller**
Improves `kronecker_solve` by replacing the existing implementation with a new one. Adds `KroneckerLinearSolver` class which holds the individual 1d matrix solvers, as well as MPI distribution information and temporary arrays. It implements:
* Ability to solve linear equations with matrices in Kronecker structure of arbitrary dimension. Supports both serial and parallel/MPI and choses the appropriate solver for each dimension (i.e. if the subcommunicator in the cartesian distribution is of size 1, then the serial solver will be used, though we have MPI).
* Per dimension, the number of calls to the underlying `DirectSolver` is reduced to 1.
* Per dimension, two `Alltoallv` calls are used to distribute the data, and each process solves a different part of the problem in the subcommunicator. (at the cost of having to re-order data a bit more, but if the communication is more expensive, then it is worth it)
* Data is re-ordered/transposed on a whole (instead of line-by-line), i.e. bulk memory copy optimizations can be used (as far as implemented in numpy).
* The `LinearSolver`, `DirectSolver` and derivative classes now also support a `transposed` flag which lets solve against the transposed matrix instead. It works also with the `KroneckerLinearSolver`.
* `kronecker_solve` still exists, but now internally instantiates a temporary `KroneckerLinearSolver` instead.

### CODE REVIEW

**Comment by yguclu** [comment]
@davschneller Thanks for your PR!

Since we use "squash commits" to merge into `devel`, the history of the individual commits will be lost together with their messages. For this reason you should provide more details in the PR's description (in the form of a list), which I will then copy and paste to the squash commit message.

**Comment by davschneller** [comment]
@yguclu Ok, thanks for letting me know; I'll try to have it done once the PR is ready for merge.

**Comment by davschneller** [comment]
So, I'm going to post that here, I guess...

**Tests on local machine**
I did do some tests on my local laptop so far (hardware not disclosed so far, OS is Linux, and numpy is from the official Arch Linux repositories), i.e. the MPI operations can be assumed weigh in rather lightly. I did solve the whole problem with 1 process (i.e. no MPI; but it did seem to use threading for LAPACK), and 8 processes (i.e. two MPI processes per sub-communicator), so the MPI cost did not really come into play. I tried to also do it with 64 MPI processes (4 MPI processes per sub-communicator) which oversubscribed the number of CPUs I had at hand (i.e. the results are most likely worse than on well-configured systems), but here MPI dominated.
I did the tests by solving linear systems with a 3-dimensional matrix, i.e. A=A_1⊗A_2⊗A_3 where all A_i have the same size which is also a power of two. Also, the matrices are tridiagonal (i.e. the solve operations should play a time-wise rather little part in the complete operation), and we use an in-place banded solver from the LAPACK pack I had at hand. The tests were each done 1000 times.

Key insights:

size (128,128,128), no MPI:
354 ms vs 153 ms (i.e. about 2,3× speedup)
(remark: the new method spends about 80% of its time in the solve operation, and uses most of the remaining time for re-ordering the dimensions and copying memory ; the old method spends about 60% in the solve method)

size (128,128,128), 8 MPI processes:
112 ms vs 40 ms (i.e. about 2,8× speedup)
(remark: the new method spent about 45% of its time in the MPI Alltoallv and the re-ordering of the blocks, and about 20% in re-ordering the dimensions and memory copying and only about 33% in the solve part ; interestingly enough, the old method spends percentually more time in the solve operations than in the Allgatherv, but still spends more time in each of these)

size (128,128,128), 64 MPI processes:
925 ms vs 37 ms (i.e. about 25× speedup; but again, this is more a synthetic test case)
(remark: the new method spent about 75% of its time in the MPI Alltoallv, the re-ordering of the blocks and dimension re-ordering took only 4% to 5%, the actual solve only about 5% of the overall time ; the old method spends about 85% in the Allgatherv part)

It should be noted that the new method, when comparing no MPI to 8 MPI processes, then the 8 MPI process variant is about 3,8× as fast. This may have several reasons, including that the no MPI version seemed to use threads for solving the right-hand sides. The 64 MPI process version cannot be compared very well, since it used a single CPU core to host about 8 processes.

**Plots (with legends placed around)**

Here is a plot of the computation times, for the old and new methods:
Figure_1
Note that the 64 MPI process lines are naturally higher than the rest.

Next, some plot which shows how much time is spent in the MPI calls **(NOTE: a lower number here can still indicate that the method takes much longer to complete than other methods; it would be more accurate to place every line of these graphs in different plots)**:
Figure_2
One could say, lower is better. Again, the 64 MPI process lines are included, but as it can be seen, in these artificial scenarios, MPI takes most of the time. The no-MPI line is also included at 0%, for completeness. The old methods seem generally to use a bit more MPI than the new versions, except that one exception at 128.

And here some plot which shows how much time is spend in the `DirectSolver`, in our case a banded solver **(NOTE: a lower number here can still indicate that the method takes much longer to complete than other methods; it would be more accurate to place every line of these graphs in different plots)**:
Figure_3
One could say, higher is better, because then our time is well-spent. Again, at 128, there is some drop to be seen. This drop also occurs in the no-MPI line (the orange one), one may assume that re-ordering the data begins to take more time here. The old method seems to spend more time in the solvers with the size increasing, but this is relativized again, when looking at the absolute performance.

**Tests on the cluster**
I also took two nodes from the CoolMUC2 cluster and used Intel Python 3.6. The interconnect was potentially a bit broken/there were problems with Intel MPI 2019. Nonetheless, the few results I got with 112 MPI processes (on two nodes) a few results; namely: the old method took too long, so I aborted the process, after the test took about 1/2h with (32,32,32). For the new method, I got results for (128,128,128). In all the cases with 112 processes, MPI takes up most of the runtime. For the old method, we spend 70% of the time in the Allgatherv, and 30% of the time in the ghost region updates. For the new method, it is the ghost region updates with about 70% (MPI Barrier), followed by about 30% for the Alltoallv in the `KroneckerLinearSolver`. Again, the interconnect might be slower here than it could be. When testing on (128,128,128) with only 2 processes (also, only the new method), the solve is again the part that takes up most of the time. Both of these runs for (128,128,128) took longer than on my local machine (546 ms for two processes, 600 ms for 112 processes).

File: psydac/linalg/kron.py [hunk 1]
```diff
@@ -325,170 +197,490 @@ def transpose(self):
...
+        """
+        sourceview = source[:self._localsize]
+        sourceview.shape = self._shapes[-1]
+
+        outslice[:] = sourceview.transpose(self._perm)
+
+    class KroneckerSolverSerialPass:
+        """
+        Solves several linear equations at the same time,
+        given that the data is already in memory.
```
Comment (yguclu) [inline psydac/linalg/kron.py:428]: This docstring should be clarified: the same equation is solved for multiple right-hand-side vectors simultaneously.

File: psydac/linalg/kron.py [hunk 2]
```diff
@@ -325,170 +197,490 @@ def transpose(self):
...
+            The MPI type of the space. Used for the Alltoallv.
+        
+        i : int
+            The index of the dimension.
+        
+        cart : CartDecomposition
+            The cartesian decomposition we use.
+
+        mglobal : int
+            The number of right-hand sizes we want to solve. Equals the product of the
```
Comment (yguclu) [inline psydac/linalg/kron.py:514]: Typo? sizes -> sides

File: psydac/linalg/kron.py [hunk 3]
```diff
@@ -325,170 +197,490 @@ def transpose(self):
...
+        Parameters
+        ----------
+        solver : DirectSolver
+            The internally used solver class.
+        
+        nglobal : int
+            The length of the dimension which we want to solve for.
+        
+        mglobal : int
+            The number of right-hand sizes we want to solve. Equals the product of the
```
Comment (yguclu) [inline psydac/linalg/kron.py:439]: Typo? sizes -> sides

**Review by yguclu — CHANGES_REQUESTED** [review]
Good job @davschneller! i have a few comments.

File: psydac/linalg/kron.py [hunk 4]
```diff
@@ -325,170 +197,490 @@ def transpose(self):
...
+            return self._datasize
+
+        def solve_pass(self, workmem, tempmem, transposed):
+            """
+            Solves the data available in workmem, assuming that all data is available locally.
+
+            Parameters
+            ----------
+            workmem : ndarray
+                The data which is used for solving. All columns to be solved are ordered contiguously.
```
[the remark this answers is not in the document]
Comment (davschneller) [reply psydac/linalg/kron.py:464]: That kind of the case. It _has_ this ordering, but `workmem` is actually passed as a one-dimensional array which has all columns written one after another contiguously in memory. It is then re-interpreted as two-dimensional data, with the size of the columns as one dimension, and all other dimensions are combined into another one (we don't care for the internal ordering of them), so, the indexing is then like `workmem[i, :]` where `i` is an index over all columns (but not specified any further). (we also convert to FORTRAN order after that all)

File: psydac/linalg/kron.py [hunk 5]
```diff
@@ -325,170 +197,490 @@ def transpose(self):
...
+        # 8 9 A B C D E F on Pr2
+        #
+
+        # NOTE: ideas for future improvements, if this is too slow:
+        #
+        # * Use MPI composite Datatypes (i.e. MPI contiguous and vector).
+        #       This may improve performance, depending on the implementation
+        #       (therefore, a library-level switch or similar would be an option here).
+        #       Mainly, we could push what happens in _blocked_to_contiguous and
+        #       _contiguous_to_blocked methods into the MPI implementation.
```
[the remark this answers is not in the document]
Comment (davschneller) [reply psydac/linalg/kron.py:565]: Yes, this is about what I thought about there. In our case, we might get away with a `Create_vector` already possibly.

File: psydac/linalg/kron.py (hunk 3 above)
Comment (davschneller) [reply psydac/linalg/kron.py]: Yes, typo. Should be fixed by now.

File: psydac/linalg/direct_solvers.py [hunk 6]
```diff
@@ -75,16 +75,23 @@ def space( self ):
...
             assert out.shape == rhs.shape
-            out[:], self._sinfo = dgbtrs(self._bmat, self._l, self._u, rhs, self._ipiv)
+
+            # support in-place operations
+            if rhs is not out:
+                rhs[:] = out
+
+            # TODO: handle non-contiguous views?
+            
+            _, self._sinfo = dgbtrs(self._bmat, self._l, self._u, rhs, self._ipiv, overwrite_b=True, trans=transposed)
```
[the remark this answers is not in the document]
Comment (davschneller) [reply psydac/linalg/direct_solvers.py]: Well yes, it was meant to be this way... It's changed now and new tests have been added.

File: psydac/linalg/kron.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (davschneller) [reply psydac/linalg/kron.py:567]: Should I add this to the current code in this PR (in addition to the existing code paths or as substitute for them)? Or shall this be done in a later PR?

**Review by yguclu — APPROVED** [review]

**Review by saidctb — APPROVED** [review]

File: psydac/linalg/kron.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (yguclu) [reply psydac/linalg/kron.py]: Thanks for the explanation.

**Pull request merged**